File: api/habits.php
<?php
/**
 * inphub: habits: manage + daily logging + streaks.
 *
 *   GET  ?action=list                      habits with today state, streaks, recent logs
 *   POST ?action=create { name, ... }
 *   POST ?action=update { id, ... }
 *   POST ?action=delete { id }
 *   POST ?action=log    { id, date?, note? }   toggle a day (unique per day)
 */

require_once __DIR__ . '/_bootstrap.php';

api_handle(function (): void {
    $uid   = current_user_id();
    $input = request_input();

    switch (action($input)) {
        case 'list': {
            $stmt = db()->prepare('SELECT * FROM habits WHERE user_id = ? ORDER BY sort_order ASC, id ASC');
            $stmt->execute([$uid]);
            $habits = $stmt->fetchAll();

            // Pull the last 140 days of logs once, group by habit.
            $logStmt = db()->prepare(
                'SELECT habit_id, logged_date, count FROM habit_logs
                 WHERE user_id = ? AND logged_date >= (CURRENT_DATE - INTERVAL 140 DAY)
                 ORDER BY logged_date ASC'
            );
            $logStmt->execute([$uid]);
            $byHabit = [];
            foreach ($logStmt->fetchAll() as $log) {
                $byHabit[(int) $log['habit_id']][] = $log;
            }

            $today = today();
            foreach ($habits as &$h) {
                $logs   = $byHabit[(int) $h['id']] ?? [];
                $target = max(1, (int) $h['target_per_period']);
                // A day counts as "done" only once the target is met, so a 3x
                // habit tapped once shows as partial instead of complete.
                $dates  = [];
                $todayCount = 0;
                foreach ($logs as $l) {
                    if ((int) $l['count'] >= $target) {
                        $dates[] = $l['logged_date'];
                    }
                    if ($l['logged_date'] === $today) {
                        $todayCount = (int) $l['count'];
                    }
                }
                $h['logs']           = $logs;
                $h['target']         = $target;
                $h['today_count']    = $todayCount;
                $h['logged_today']   = $todayCount >= $target;
                [$cur, $best]        = streaks($dates, (string) $h['frequency']);
                $h['current_streak'] = $cur;
                $h['best_streak']    = $best;
            }
            unset($h);
            ok($habits);
            break;
        }

        case 'create': {
            $name = str_or_null(input_get($input, 'name'));
            if ($name === null) {
                fail('Habit name is required.', 422);
            }
            $stmt = db()->prepare(
                'INSERT INTO habits (user_id, name, description, frequency, target_per_period, color, icon, sort_order)
                 VALUES (?, ?, ?, ?, ?, ?, ?, ?)'
            );
            $stmt->execute([
                $uid,
                $name,
                str_or_null(input_get($input, 'description')),
                valid_enum(input_get($input, 'frequency'), ['daily', 'weekly'], 'daily'),
                max(1, (int) input_get($input, 'target_per_period', 1)),
                str_or_null(input_get($input, 'color')) ?? '#4f8cff',
                str_or_null(input_get($input, 'icon')),
                (int) input_get($input, 'sort_order', 0),
            ]);
            ok(['id' => (int) db()->lastInsertId()]);
            break;
        }

        case 'update': {
            $h = fetch_owned('habits', (int) input_get($input, 'id'), $uid);
            $stmt = db()->prepare(
                'UPDATE habits SET name=?, description=?, frequency=?, target_per_period=?, color=?, icon=?, is_active=?, sort_order=?
                 WHERE id=? AND user_id=?'
            );
            $stmt->execute([
                str_or_null(input_get($input, 'name', $h['name'])) ?? $h['name'],
                array_key_exists('description', $input) ? str_or_null($input['description']) : $h['description'],
                valid_enum(input_get($input, 'frequency', $h['frequency']), ['daily', 'weekly'], $h['frequency']),
                max(1, (int) input_get($input, 'target_per_period', $h['target_per_period'])),
                str_or_null(input_get($input, 'color', $h['color'])) ?? $h['color'],
                array_key_exists('icon', $input) ? str_or_null($input['icon']) : $h['icon'],
                (int) (bool) input_get($input, 'is_active', $h['is_active']),
                (int) input_get($input, 'sort_order', $h['sort_order']),
                (int) $h['id'],
                $uid,
            ]);
            ok(['id' => (int) $h['id']]);
            break;
        }

        case 'delete': {
            $h = fetch_owned('habits', (int) input_get($input, 'id'), $uid);
            $del = db()->prepare('DELETE FROM habits WHERE id=? AND user_id=?');
            $del->execute([(int) $h['id'], $uid]);
            ok(['deleted' => (int) $h['id']]);
            break;
        }

        case 'log': {
            $h      = fetch_owned('habits', (int) input_get($input, 'id'), $uid);
            $date   = str_or_null(input_get($input, 'date')) ?? today();
            $target = max(1, (int) $h['target_per_period']);

            // habit_logs.count is what makes target_per_period mean anything:
            // the unique key is (habit_id, logged_date), so a habit you want to
            // do 3x a day counts up within the single row for that day rather
            // than inserting three. Each tap advances the count; the tap after
            // the target is reached clears the day (so it stays a toggle).
            $check = db()->prepare('SELECT id, count FROM habit_logs WHERE habit_id=? AND logged_date=?');
            $check->execute([(int) $h['id'], $date]);
            $row = $check->fetch();

            if ($row === false) {
                $ins = db()->prepare(
                    'INSERT INTO habit_logs (user_id, habit_id, logged_date, count, note) VALUES (?, ?, ?, 1, ?)'
                );
                $ins->execute([$uid, (int) $h['id'], $date, str_or_null(input_get($input, 'note'))]);
                log_activity($uid, 'habit.logged', 'habit', (int) $h['id'],
                    'Logged habit: ' . $h['name'] . ($target > 1 ? " (1/$target)" : ''), 'user');
                ok(['logged' => true, 'count' => 1, 'target' => $target, 'date' => $date]);
            } elseif ((int) $row['count'] < $target) {
                $next = (int) $row['count'] + 1;
                $upd = db()->prepare('UPDATE habit_logs SET count=? WHERE id=?');
                $upd->execute([$next, (int) $row['id']]);
                log_activity($uid, 'habit.logged', 'habit', (int) $h['id'],
                    'Logged habit: ' . $h['name'] . " ($next/$target)", 'user');
                ok(['logged' => true, 'count' => $next, 'target' => $target, 'date' => $date]);
            } else {
                $del = db()->prepare('DELETE FROM habit_logs WHERE id=?');
                $del->execute([(int) $row['id']]);
                log_activity($uid, 'habit.unlogged', 'habit', (int) $h['id'],
                    'Cleared habit: ' . $h['name'] . ' on ' . $date, 'user');
                ok(['logged' => false, 'count' => 0, 'target' => $target, 'date' => $date]);
            }
            break;
        }

        default:
            fail('Unknown action.', 404);
    }
});

/**
 * Given an ascending list of YYYY-MM-DD strings, return [currentStreak, bestStreak].
 *
 * $frequency 'daily' counts consecutive days; 'weekly' counts consecutive
 * ISO weeks with at least one log. Without the weekly branch a once-a-week
 * habit read "current 1 / best 1" forever, which made the whole frequency
 * setting decorative.
 */
function streaks(array $dates, string $frequency = 'daily'): array
{
    if (count($dates) === 0) {
        return [0, 0];
    }
    if ($frequency === 'weekly') {
        return week_streaks($dates);
    }
    $set = array_flip($dates);

    // Best streak: longest consecutive run anywhere.
    $best = 0;
    foreach (array_keys($set) as $d) {
        $prev = date('Y-m-d', strtotime($d . ' -1 day'));
        if (!isset($set[$prev])) {
            // Start of a run, count forward.
            $len = 1;
            $cur = $d;
            while (isset($set[date('Y-m-d', strtotime($cur . ' +1 day'))])) {
                $cur = date('Y-m-d', strtotime($cur . ' +1 day'));
                $len++;
            }
            $best = max($best, $len);
        }
    }

    // Current streak: from today (or yesterday if today not yet logged) backwards.
    $current = 0;
    $cursor = today();
    if (!isset($set[$cursor])) {
        $cursor = date('Y-m-d', strtotime('-1 day'));
    }
    while (isset($set[$cursor])) {
        $current++;
        $cursor = date('Y-m-d', strtotime($cursor . ' -1 day'));
    }

    return [$current, $best];
}

/** Consecutive-ISO-week runs, for habits with frequency = 'weekly'. */
function week_streaks(array $dates): array
{
    $weeks = [];
    foreach ($dates as $d) {
        $weeks[date('o-W', strtotime($d))] = true;
    }

    $best = 0;
    foreach (array_keys($weeks) as $w) {
        $prev = shift_week($w, -1);
        if (isset($weeks[$prev])) {
            continue;   // not the start of a run
        }
        $len = 1;
        $cur = $w;
        while (isset($weeks[shift_week($cur, 1)])) {
            $cur = shift_week($cur, 1);
            $len++;
        }
        $best = max($best, $len);
    }

    $current = 0;
    $cursor  = date('o-W');
    if (!isset($weeks[$cursor])) {
        $cursor = shift_week($cursor, -1);   // this week not logged yet
    }
    while (isset($weeks[$cursor])) {
        $current++;
        $cursor = shift_week($cursor, -1);
    }
    return [$current, $best];
}

/** Move an "o-W" ISO week key by $delta weeks. */
function shift_week(string $key, int $delta): string
{
    [$year, $week] = array_map('intval', explode('-', $key));
    $ts = strtotime(sprintf('%04dW%02d', $year, $week) . ' ' . ($delta >= 0 ? '+' : '-') . abs($delta) . ' week');
    return date('o-W', $ts ?: time());
}

File: server/Endpoints/Habits.cs
using System.Globalization;
using Inphub.Core;

namespace Inphub.Endpoints;

public static class Habits
{
    static readonly string[] Frequencies = ["daily", "weekly"];

    // GET/POST /api/habits: list (with streaks), create, update, delete, log.
    public static void Map(WebApplication app) =>
        app.MapMethods("/api/habits", ["GET", "POST"], Api.Handle(Run));

    // Dispatches on ?action=.
    static object? Run(Req req)
    {
        var uid = req.Uid;
        var input = req.Input;

        switch (req.Action)
        {
            case "list":
                return List(uid);
            case "create":
            {
                var name = input.Str("name") ?? throw Api.Fail("Habit name is required.", 422);
                var id = Db.Insert(
                    @"INSERT INTO habits (user_id, name, description, frequency, target_per_period, color, icon, sort_order)
                      VALUES (@uid, @name, @description, @frequency, @target, @color, @icon, @sort_order)",
                    new
                    {
                        uid, name,
                        description = input.Str("description"),
                        frequency = Api.ValidEnum(input.Text("frequency"), Frequencies, "daily"),
                        target = Math.Max(1, input.Int("target_per_period", 1)),
                        color = input.Str("color") ?? "#4f8cff",
                        icon = input.Str("icon"),
                        sort_order = input.Int("sort_order"),
                    });
                return new { id };
            }
            case "update":
            {
                var habit = Api.FetchOwned("habits", input.Int("id"), uid);
                var id = Convert.ToInt32(habit["id"]);
                var frequency = (string)habit["frequency"]!;
                Db.Execute(
                    @"UPDATE habits SET name = @name, description = @description, frequency = @frequency, target_per_period = @target,
                        color = @color, icon = @icon, is_active = @is_active, sort_order = @sort_order
                      WHERE id = @id AND user_id = @uid",
                    new
                    {
                        id, uid,
                        name = input.Str("name") ?? (string)habit["name"]!,
                        description = input.StrIfSent("description", habit["description"]),
                        frequency = Api.ValidEnum(input.Str("frequency") ?? frequency, Frequencies, frequency),
                        target = Math.Max(1, input.Int("target_per_period", Convert.ToInt32(habit["target_per_period"]))),
                        color = input.Str("color") ?? (string)habit["color"]!,
                        icon = input.StrIfSent("icon", habit["icon"]),
                        is_active = input.Bool("is_active", Convert.ToInt32(habit["is_active"]) != 0) ? 1 : 0,
                        sort_order = input.Int("sort_order", Convert.ToInt32(habit["sort_order"])),
                    });
                return new { id };
            }
            case "delete":
            {
                var habit = Api.FetchOwned("habits", input.Int("id"), uid);
                var id = Convert.ToInt32(habit["id"]);
                Db.Execute("DELETE FROM habits WHERE id = @id AND user_id = @uid", new { id, uid });
                return new { deleted = id };
            }
            case "log":
                return Log(uid, input);
            default:
                throw Api.Fail("Unknown action.", 404);
        }
    }

    // Every habit with its recent logs, today's count and streaks.
    static object List(int uid)
    {
        var habits = Db.Rows("SELECT * FROM habits WHERE user_id = @uid ORDER BY sort_order ASC, id ASC", new { uid });
        var logs = Db.Rows(
            @"SELECT habit_id, logged_date, count FROM habit_logs
              WHERE user_id = @uid AND logged_date >= (CURRENT_DATE - INTERVAL 140 DAY)
              ORDER BY logged_date ASC",
            new { uid });

        var today = AppInfo.Today();
        foreach (var habit in habits)
        {
            var id = Convert.ToInt32(habit["id"]);
            var target = Math.Max(1, Convert.ToInt32(habit["target_per_period"]));
            var own = logs.Where(l => Convert.ToInt32(l["habit_id"]) == id).ToList();

            var doneDates = new List<string>();
            var todayCount = 0;
            foreach (var log in own)
            {
                var date = (string)log["logged_date"]!;
                var count = Convert.ToInt32(log["count"]);
                if (count >= target) doneDates.Add(date);
                if (date == today) todayCount = count;
            }

            var (current, best) = habit["frequency"] as string == "weekly" ? WeekStreaks(doneDates) : DayStreaks(doneDates);
            habit["logs"] = own;
            habit["target"] = target;
            habit["today_count"] = todayCount;
            habit["logged_today"] = todayCount >= target;
            habit["current_streak"] = current;
            habit["best_streak"] = best;
        }
        return habits;
    }

    // One tap: counts the day up toward the target; the tap after the target clears it.
    static object Log(int uid, Input input)
    {
        var habit = Api.FetchOwned("habits", input.Int("id"), uid);
        var id = Convert.ToInt32(habit["id"]);
        var name = (string)habit["name"]!;
        var date = input.Str("date") ?? AppInfo.Today();
        var target = Math.Max(1, Convert.ToInt32(habit["target_per_period"]));

        var row = Db.Row("SELECT id, count FROM habit_logs WHERE habit_id = @id AND logged_date = @date", new { id, date });
        if (row == null)
        {
            Db.Execute("INSERT INTO habit_logs (user_id, habit_id, logged_date, count, note) VALUES (@uid, @id, @date, 1, @note)",
                new { uid, id, date, note = input.Str("note") });
            Activity.Log(uid, "habit.logged", "habit", id, "Logged habit: " + name + (target > 1 ? $" (1/{target})" : ""));
            return new { logged = true, count = 1, target, date };
        }

        var count = Convert.ToInt32(row["count"]);
        if (count < target)
        {
            var next = count + 1;
            Db.Execute("UPDATE habit_logs SET count = @next WHERE id = @logId", new { next, logId = row["id"] });
            Activity.Log(uid, "habit.logged", "habit", id, $"Logged habit: {name} ({next}/{target})");
            return new { logged = true, count = next, target, date };
        }

        Db.Execute("DELETE FROM habit_logs WHERE id = @logId", new { logId = row["id"] });
        Activity.Log(uid, "habit.unlogged", "habit", id, $"Cleared habit: {name} on {date}");
        return new { logged = false, count = 0, target, date };
    }

    // Current and best runs of consecutive days.
    static (int Current, int Best) DayStreaks(List<string> dates)
    {
        var days = dates.Select(d => DateTime.ParseExact(d, "yyyy-MM-dd", CultureInfo.InvariantCulture)).ToHashSet();
        return Runs(days, DateTime.Today, d => d.AddDays(1), d => d.AddDays(-1));
    }

    // Current and best runs of consecutive ISO weeks with at least one done day.
    static (int Current, int Best) WeekStreaks(List<string> dates)
    {
        var weeks = dates.Select(d => WeekStart(DateTime.ParseExact(d, "yyyy-MM-dd", CultureInfo.InvariantCulture))).ToHashSet();
        return Runs(weeks, WeekStart(DateTime.Today), w => w.AddDays(7), w => w.AddDays(-7));
    }

    // Longest run anywhere, and the run ending now (or one step back if now isn't done yet).
    static (int Current, int Best) Runs(HashSet<DateTime> set, DateTime now, Func<DateTime, DateTime> next, Func<DateTime, DateTime> prev)
    {
        var best = 0;
        foreach (var start in set)
        {
            if (set.Contains(prev(start))) continue;
            var length = 1;
            var cursor = start;
            while (set.Contains(next(cursor)))
            {
                cursor = next(cursor);
                length++;
            }
            best = Math.Max(best, length);
        }

        var current = 0;
        var at = set.Contains(now) ? now : prev(now);
        while (set.Contains(at))
        {
            current++;
            at = prev(at);
        }
        return (current, best);
    }

    // The Monday that starts a date's ISO week.
    static DateTime WeekStart(DateTime d) => d.Date.AddDays(-(((int)d.DayOfWeek + 6) % 7));
}
